cloudinit/sources/DataSourceScaleway.py

```diff
@@ -177,51 +186,152 @@ def __init__(self, sys_cfg, distro, paths):
+                LOG.debug("%s is reachable", url)
+                self.metadata_url = f"{url}/conf?format=json"
+                self.userdata_url = f"{url}/user_data/cloud-init"
+                self.vendordata_url = f"{url}/vendor_data/cloud-init"
+                return
+        LOG.debug("Unable to reach api-metadata at %s", urls)
+        raise ConnectionError
 
     def _crawl_metadata(self):
         resp = url_helper.readurl(
-            self.metadata_address, timeout=self.timeout, retries=self.retries
+            self.metadata_url, timeout=self.timeout, retries=self.retries
         )
         self.metadata = json.loads(util.decode_binary(resp.contents))
 
         self.userdata_raw = query_data_api(
-            "user-data", self.userdata_address, self.retries, self.timeout
+            "user-data", self.userdata_url, self.retries, self.timeout
         )
         self.vendordata_raw = query_data_api(
-            "vendor-data", self.vendordata_address, self.retries, self.timeout
+            "vendor-data", self.vendordata_url, self.retries, self.timeout
         )
 
+    def _set_urls_on_ip_version(self, proto, urls):
+
+        if proto not in ["ipv4", "ipv6"]:
+            LOG.debug("Invalid IP version : %s", proto)
+            return []
+
+        filtered_urls = []
+        for url in urls:
+            # Numeric IPs
+            address = urlparse(url).netloc
+            if address[0] == "[":
+                address = address[1:-1]
+            addr_proto = socket.getaddrinfo(
+                address, None, proto=socket.IPPROTO_TCP
+            )[0][0]
+            if addr_proto == socket.AF_INET and proto == "ipv4":
+                filtered_urls += [url]
+                continue
+            elif addr_proto == socket.AF_INET6 and proto == "ipv6":
+                filtered_urls += [url]
+                continue
+
+        return filtered_urls
+
     def _get_data(self):
         if not on_scaleway():
             return False
 
         if self._fallback_interface is None:
             self._fallback_interface = net.find_fallback_nic()
-        try:
-            with EphemeralDHCPv4(
-                self._fallback_interface,
-                tmp_dir=self.distro.get_tmp_exec_path(),
-            ):
-                util.log_time(
-                    logfunc=LOG.debug,
-                    msg="Crawl of metadata service",
-                    func=self._crawl_metadata,
-                )
-        except (NoDHCPLeaseError) as e:
-            util.logexc(LOG, str(e))
-            return False
+        if self.has_ipv4:
+            try:
+                # DHCPv4 waits for timeout defined in /etc/dhcp/dhclient.conf
+                # before giving up. Lower it in config file and try it first as
+                # it will only reach timeout on VMs with only IPv6 addresses.
+                with EphemeralDHCPv4(
+                    self._fallback_interface,
+                    tmp_dir=self.distro.get_tmp_exec_path(),
```

Comment:
This line can be dropped. `tmp_dir` isn't actually used anymore - that kwarg needs to get removed from the definition

Reply:
Removed.